docs/sequence.rst

```diff
@@ -0,0 +1,672 @@
+
+
+Importing Sequence
+------------------
+
+When using ``Sequence`` in a Haskell source file you should always use a
+``qualified`` import becasue it exports names that clash with the standard
+Prelude (you can import the type constructor and some operators on their own
+though!).
+
+::
+
+    import Data.Sequence (Seq(..), (<|), (|>), (><))
+    import qualified Data.Sequence as Seq
+
+
+Common API Functions
+--------------------
+
+.. NOTE::
+   ``fromList [some,sequence,elements]`` is how a ``Seq`` is printed.
+
+Construction and Conversion
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Create an empty sequence
+""""""""""""""""""""""""
+
+::
+
+    Seq.empty :: Seq a
+    Seq.empty = ...
+
+:seq:`empty` creates a sequence with zero elements.
+
+::
+
+    Seq.empty
+    > fromList []
+
+
+Create a sequence with one element (singleton)
+""""""""""""""""""""""""""""""""""""""""""""""
+
+::
+
+    Seq.singleton :: a -> Seq a
+    Seq.singleton x = ...
+
+:seq:`singleton` creates a sequence with the single element ``x`` in it.
+
+::
+
+    Seq.singleton "containers"
+    > fromList ["containers"]
+
+    Seq.singleton 1
+    > fromList [1]
+
+Create a sequence with the same element repeated
+""""""""""""""""""""""""""""""""""""""""""""""""
+
+::
+
+    Seq.replicate :: Int -> a -> Seq a
+    Seq.replicate n x = ...
+
+:seq:`replicate` creates a sequence with same element ``x`` repeated ``n``
+times.
+
+::
+
+    Seq.replicate 0 "hi"
+    > fromList []
+
+    Seq.replicate 3 "hi"
+    > fromList ["hi","hi","hi"]
+
+Create a sequence from a list
+"""""""""""""""""""""""""""""
+
+::
+
+    Seq.fromList :: [a] -> Seq a
+    Seq.FromList xs = ...
+
+:seq:`fromList` creates a sequence containing the elements of the list
+``xs``. Sequences allow duplicate so all elements will be included in the order
+given.
+
+::
+
+    Seq.fromList ["base", "containers", "QuickCheck"]
+    > fromList ["base","containers","QuickCheck"]
+
+    Seq.fromList [0, 1, 1, 2, 3, 1]
+    > fromList [0,1,1,2,3,1]
+
+Adding to an existing sequence
+""""""""""""""""""""""""""""""
+
+::
+
+    (<|) :: a -> Seq a -> Seq a
+    x <| xs = ...
+
+    (|>) :: Seq a -> a -> Seq a
+    xs |> x = ...
+
+    (><) :: Seq a -> Seq a -> Seq a
+    l >< r = ...
+
+- ``x <| xs`` places the element ``x`` at the beginning of the sequence ``xs``..
+
+- ``xs |> x`` places the element ``x`` at the end of the sequence ``xs``.
+
+- ``l >< r`` combines the two sequences ``l`` and ``r`` together.
+
+
+Create a list from a sequence
+"""""""""""""""""""""""""""""
+
+::
+
+    foldr (:) [] :: Seq a -> [a]
+
+
+There is no ``toList`` function in the Sequence module since it can be
+`easily implemented <https://wiki.haskell.org/Foldable_and_Traversable>`_ with a
+fold using ``Seq``'s `Foldable
+<https://wiki.haskell.org/Typeclassopedia#Foldable>`_ instance.
+
+::
+
+    foldr (:) [] (Seq.fromList ["base", "containers", "QuickCheck"])
+    > ["base","containers","QuickCheck"]
+
+
+Pattern Matching
```

Comment:
The patterns are bidirectional. So for appropriate `containers` *and GHC* versions, these patterns can completely replace `<|`, `|>`, `empty`, `null` (except where that's actually desired), `viewl`, and `viewr`. I tend to think that in a tutorial like this, the patterns should be presented as the primary way to perform these operations. Only people concerned about Haskell 2010 compliance or older versions of GHC need to deal with the rest. The only annoying bit is that `:<|` and `:|>` are a bit verbose for historical reasons. If we were building from scratch, we might've used `:<` and `:>`, but those were taken for inferior view constructors.

Reply:
Is it reasonable to assume that everyone would be on the "appropriate `containers` and GHC versions"? I'm assuming they would need containers >= 0.5.10 and ghc >= 8.2 which not everyone is on at this point (especially if they're using a system GHC).

If they need something that new I would be hesitant to put these as the primary way of doing these operations. I don't even use 8.2 regularly right now because ghc-mod still doesn't support 8.2 so editor tooling is kinda borked.

How would you feel about revisiting this in a few months when all the 8.2 kinks are worked out?

Reply:
GHC 8.0 will do, unless you want to run `-Wall` or `-Wincomplete-patterns`. It's up to you; I just think the pattern synonyms make sequences easier to work with and reduce the mental clutter of silly view functions.